### PR TITLE
mon, client: Filter statfs for file systems with a single data pool

### PR DESCRIPTION
**Pull request by fullerdj**

### CODE REVIEW

**Comment by fullerdj** [comment]
@batrick @jcsp Can we merge this?

**Comment by batrick** [comment]
@fullerdj sorry I msised this because of no cephfs label. I'll take a look now.

**Comment by fullerdj** [comment]
looks like it needs a rebase while I'm at it.

File: src/messages/MStatfs.h [hunk 1]
```diff
@@ -35,17 +37,24 @@ class MStatfs : public PaxosServiceMessage {
...
   void print(ostream& out) const override {
-    out << "statfs(" << get_tid() << " v" << version << ")";
+    out << "statfs(" << get_tid() << " pool "
+        << (data_pool ? *data_pool : -1) << " v" << version << ")";
   }
 
   void encode_payload(uint64_t features) override {
     paxos_encode();
     ::encode(fsid, payload);
+    ::encode(data_pool, payload);
```
Comment (batrick) [inline src/messages/MStatfs.h]: Is this always encoded as an int?

File: src/mon/MgrStatMonitor.cc [hunk 2]
```diff
@@ -290,11 +291,11 @@ bool MgrStatMonitor::preprocess_getpoolstats(MonOpRequestRef op)
   return true;
 }
 
-bool MgrStatMonitor::preprocess_statfs(MonOpRequestRef op)
-{
+bool MgrStatMonitor::preprocess_statfs(MonOpRequestRef op) {
```
Comment (batrick) [inline src/mon/MgrStatMonitor.cc]: Unnecessary code churn (and conflicts with style of the file).

File: src/mon/PGMap.h [hunk 3]
```diff
@@ -147,6 +147,43 @@ class PGMapDigest {
   void pool_cache_io_rate_summary(Formatter *f, ostream *out,
                                   uint64_t poolid) const;
 
+  /**
+   * Return the number of additional bytes that can be stored in this
+   * pool before the first OSD fills up, accounting for PG overhead.
+   */
+  int64_t get_pool_free_space(const OSDMap &osd_map, int poolid) const;
+
+  float pool_raw_used_rate(const OSDMap &osd_map, int poolid) const
```
Comment (batrick) [inline src/mon/PGMap.h]: Why not put this as a static function in PGMap.cc?

File: src/messages/MStatfs.h (hunk 1 above)
Comment (fullerdj) [inline src/messages/MStatfs.h]: It's encoded as a boost::optional<int>. There's a defined encoding for that:

https://github.com/ceph/ceph/blob/master/src/include/encoding.h#L289

File: src/mon/MgrStatMonitor.cc (hunk 2 above)
Comment (fullerdj) [inline src/mon/MgrStatMonitor.cc]: More likely unnecessary press of the shift key whilst scrolling.

File: src/mon/PGMap.h (hunk 3 above)
Comment (fullerdj) [inline src/mon/PGMap.h]: Some of these stats gathering functions are rather new; we seem to be interested in exposing more statistics that rely on telling the user how many more bytes they're allowed to store.

 It's certainly only used in PGMap.cc for now, though, so I'll move it there.

**Comment by fullerdj** [comment]
For a test, we can just see if the report accounts for replication/ec overhead. I'll come up with something.

File: src/messages/MStatfs.h (hunk 1 above)
Comment (batrick) [inline src/messages/MStatfs.h]: Ah, cool.

File: src/mon/PGMap.h (hunk 3 above)
Comment (batrick) [inline src/mon/PGMap.h]: Let's just keep it in PGMap.cc for now.

File: src/messages/MStatfs.h [hunk 4]
```diff
@@ -22,10 +22,12 @@
 class MStatfs : public PaxosServiceMessage {
 public:
   uuid_d fsid;
+  boost::optional<int> data_pool;
 
-  MStatfs() : PaxosServiceMessage(CEPH_MSG_STATFS, 0) {}
-  MStatfs(const uuid_d& f, ceph_tid_t t, version_t v) :
-    PaxosServiceMessage(CEPH_MSG_STATFS, v), fsid(f) {
+  MStatfs() : PaxosServiceMessage(CEPH_MSG_STATFS, 0, 2) {}
```
Comment (liewegas) [inline src/messages/MStatfs.h]: should use the const definitions at the top of the class like other messages... see e.g.
https://github.com/ceph/ceph/blob/master/src/messages/MOSDOp.h#L36

File: src/messages/MStatfs.h [hunk 5]
```diff
@@ -22,10 +22,12 @@
 class MStatfs : public PaxosServiceMessage {
 public:
   uuid_d fsid;
+  boost::optional<int> data_pool;
```
Comment (liewegas) [inline src/messages/MStatfs.h]: this should be int64_t instead of int

File: src/mon/MgrStatMonitor.cc [hunk 6]
```diff
@@ -25,8 +25,9 @@ class MgrPGStatService : public MonPGStatService {
     return nullptr;
   }
 
-  ceph_statfs get_statfs() const override {
-    return digest.get_statfs();
+  ceph_statfs get_statfs(OSDMap& osdmap,
+			 boost::optional<int> data_pool) const override {
```
Comment (liewegas) [inline src/mon/MgrStatMonitor.cc]: s/int/int64_t/

File: src/mon/PGMap.cc [hunk 7]
```diff
@@ -614,6 +614,21 @@ void PGMapDigest::pool_cache_io_rate_summary(Formatter *f, ostream *out,
   cache_io_rate_summary(f, out, p->second.first, ts->second);
 }
 
+int64_t PGMapDigest::get_pool_free_space(const OSDMap &osd_map,
+					 int poolid) const
```
Comment (liewegas) [inline src/mon/PGMap.cc]: int64_t

File: src/mon/PGMap.h [hunk 8]
```diff
@@ -171,13 +208,24 @@ class PGMapDigest {
...
+  ceph_statfs get_statfs(OSDMap &osdmap, boost::optional<int> data_pool) const {
     ceph_statfs statfs;
-    // these are in KB.
-    statfs.kb = osd_sum.kb;
-    statfs.kb_used = osd_sum.kb_used;
-    statfs.kb_avail = osd_sum.kb_avail;
-    statfs.num_objects = pg_sum.stats.sum.num_objects;
+
+    if (data_pool) {
+      const pool_stat_t &stat = pg_pool_sum.at(*data_pool);
```
Comment (liewegas) [inline src/mon/PGMap.h]: this should probably fall back if the pool doesn't exist (e.g., if a client statfs request races with a pool deletion)

**Comment by batrick** [comment]
Tracker ticket: http://tracker.ceph.com/issues/19109

**Comment by batrick** [comment]
@fullerdj your rebase got screwed up somehow.

**Comment by fullerdj** [comment]
It seems so. I'm writing up a test, so at least this will make sure this PR doesn't get merged beforehand.

**Comment by fullerdj** [comment]
Ok, I've added a test. This should be ready to go.

File: qa/tasks/cephfs/test_misc.py [hunk 9]
```diff
@@ -130,3 +131,21 @@ def test_evict_client(self):
...
+        j = json.loads(out)
+        pool_size = int(j['size'])
+
+        proc = self.mount_a.run_shell(['df', '.'])
+        output = proc.stdout.getvalue()
+        fs_avail = output.split('\n')[1].split()[3]
+        fs_avail = float(fs_avail) * 1024
+
+        ratio = (raw_avail / pool_size) / fs_avail
+        assert ratio > 0.9 and ratio < 1.1
```
Comment (tchaikov) [inline qa/tasks/cephfs/test_misc.py]: nit, could just put
```py
assert 0.9 < ratio < 1.1
```

**Comment by batrick** [comment]
````
2017-08-02T21:07:04.939 INFO:tasks.cephfs_test_runner:
2017-08-02T21:07:04.939 INFO:tasks.cephfs_test_runner:======================================================================
2017-08-02T21:07:04.939 INFO:tasks.cephfs_test_runner:FAIL: test_lifecycle (tasks.cephfs.test_volume_client.TestVolumeClient)
2017-08-02T21:07:04.939 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2017-08-02T21:07:04.940 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2017-08-02T21:07:04.940 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-pdonnell-testing-20170802/qa/tasks/cephfs/test_volume_client.py", line 239, in test_lifecycle
2017-08-02T21:07:04.940 INFO:tasks.cephfs_test_runner:    (data_bin_mb + other_bin_mb) * 1024 * 1024)
2017-08-02T21:07:04.940 INFO:tasks.cephfs_test_runner:AssertionError: 0 not greater than 10485760
2017-08-02T21:07:04.940 INFO:tasks.cephfs_test_runner:
2017-08-02T21:07:04.940 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2017-08-02T21:07:04.940 INFO:tasks.cephfs_test_runner:Ran 7 tests in 1142.364s
````

From /ceph/teuthology-archive/pdonnell-2017-08-02_17:25:29-fs-wip-pdonnell-testing-20170802-distro-basic-smithi/1474760/teuthology.log

I'm checking if there's an easy fix for that but maybe you will find it first @fullerdj.

**Comment by batrick** [comment]
@fullerdj this needs a rebase.

**Comment by fullerdj** [comment]
> I'm checking if there's an easy fix for that but maybe you will find it first @fullerdj.

The test is racy. Even though it does conv=fdatasync, that is a nop on fuse. This probably used to work because of the previous dd; although it looks for ``data_bin_mb + other_bin_mb`` bytes, but df would have reported ``data_bin_mb * pool_size`` bytes used. I'll fix the test.

**Comment by liewegas** [comment]
rados suite looks good: http://pulpito.ceph.com/sage-2017-08-04_23:16:58-rados-wip-sage-testing-20170804b-distro-basic-smithi/

**Pull request merged**